compiler/rustc_hir_analysis/src/check/region.rs

```diff
@@ -90,6 +94,12 @@ fn resolve_block<'tcx>(visitor: &mut RegionResolutionVisitor<'tcx>, blk: &'tcx h
 
     let prev_cx = visitor.cx;
 
+    if let hir::BlockCheckMode::UnsafeBlock(src) = blk.rules
```

Comment:
This should not check for unsafe blocks. It should just check for references to static mut everywhere.

I have no idea what is the best place to introduce such a check. `region.rs` sounds like the wrong place since this check has nothing to do with regions. Can't you just add a new lint pass that walks the HIR?

Reply:
Outside of unsafe block it's already an error.
```rust
fn main() {
    static mut X: i32 = 1;
    let _y = &X;
}
```
error:
```
error[E0133]: use of mutable static is unsafe and requires unsafe function or block
 --> main.rs:3:14
  |
3 |     let _y = &X;
  |              ^^ use of mutable static
  |
  = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error: aborting due to previous error

For more information about this error, try `rustc --explain E0133`.
```

I'm not sure where the best place to move this check is. I think @compiler-errors could help by indicating where it should be located.

Reply:
> Outside of unsafe block it's already an error.

Yes I know. But The fact that they are unsafe and the fact that we want to deprecate them are orthogonal enough that I don't think the lint should care about this fact.

Reply:
For lint, I think you mean this line of code: https://github.com/rust-lang/rust/pull/117556/files#diff-61733c7ed1e34338fa6f251e6f248dc1a6ebae1f828fa955b6d862184222c103R61

Reply:
Yes.

Wait why do you have this check twice? In the HIR crate and as a lint? That shouldn't be.

Reply:
I have to do that; this is how lint works.

The code for lint should be in a different place than HIR hard errors, as far as I know.

Reply:
No, sorry -- code duplication is never something you have to do.

We'll want a single implementation of this pass, which emits the lint. Then we have edition-dependent levels (warn/deny) for the lint. That's how such lints are usually handled, I think.

I don't know how we usually handle diagnostics which are a hard error in some edition and a warning in another -- do we have an existing diagnostic like that? But we *definitely* don't want two implementations of the same check. That's just bad for maintenance.

Reply:
> I don't know how we usually handle diagnostics which are a hard error in some edition and a warning in another -- do we have an existing diagnostic like that?

We do, for bare trait objects. I have linked to it [in my comment above](https://github.com/rust-lang/rust/pull/117556#issuecomment-1793281319) where I originally suggested adding the error message. There is a function that checks whether there is a bare trait, and if it finds one, it either emits a lint or an error, depending on the edition. We do need to tweak the wording between the two, not sure if two or one ftl messages would be better. In general, I think we should copy the approach we have for bare trait objects.

The intent of my comment was to make sure that we don't make any implementation choices that make sharing of code between error and lint hard. I had the suspicion that like the PR was originally written, via a traditional lint pass, we'd likely have to move it once we do the error. So as we want to have both, I suggested to just do it here in this PR.

The way the PR implements the error and the lint currently is by *not* sharing code, which is what I wanted to avoid (and I agree with you that it's not needed).

Reply:
So, TLDR, I agree with @RalfJung that there should be one check only, with a dispatch later depending on the edition.

Reply:
I agree with the above and would recommend looking at the bare trait object lint/error that @est31 has linked as an example.